### PR TITLE
chore: upgrade to reqwest 0.12.4 and rustls 0.22

### LINKED ISSUE #23370: Update to Reqwest 0.12
So currently Reqwest is locked to a specific version. https://github.com/denoland/deno/blob/main/Cargo.toml#L148

For that reason, I am unable to compile deno-runtime in one of my projects as I have another library that requires an updated version. 

This PR https://github.com/seanmonstar/reqwest/pull/2185 was merged fixing the Mac OS bug but was released in 0.12. 
It would also be nice if we could get Hyper Updated and HTTP Updated and remove the locked package renames. But that is going to require a lot of refactoring from what I can tell.

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit updates Deno to use `reqwest` at 0.12.4
and `rustls` at 0.22. Other related crates were updated
as well to match versions accepted by `reqwest` and `rustls`.

Note: we are not using the latest available `rustls` yet,
but this upgrade was non-trivial already, so a bump to
0.23 for `rustls` will be done in a separate commit.

Closes #23370

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1890,7 +1893,7 @@ dependencies = [
  "deno_core",
  "deno_native_certs",
  "rustls",
- "rustls-pemfile",
+ "rustls-pemfile 2.1.2",
```
Comment (bartlomieju) [inline Cargo.lock:1896]: Check if we can only use a single version

File: Cargo.lock [hunk 2]
```diff
@@ -3549,7 +3554,7 @@ dependencies = [
  "socket2",
  "widestring",
  "windows-sys 0.48.0",
- "winreg",
+ "winreg 0.50.0",
```
Comment (bartlomieju) [inline Cargo.lock:3557]: Check if we can use only a single version

File: Cargo.lock [hunk 3]
```diff
@@ -5462,7 +5472,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "a9aace74cb666635c918e9c12bc0d348266037aa8eb599b5cba565709a8dff00"
 dependencies = [
  "openssl-probe",
- "rustls-pemfile",
+ "rustls-pemfile 1.0.4",
```
Comment (bartlomieju) [inline Cargo.lock:5475]: Ditto

File: Cargo.toml [hunk 4]
```diff
@@ -369,3 +368,8 @@ opt-level = 3
 opt-level = 3
 [profile.release.package.base64-simd]
 opt-level = 3
+
+[patch.crates-io]
+denokv_proto = { git = "https://github.com/denoland/denokv", branch = "main" }
+denokv_remote = { git = "https://github.com/denoland/denokv", branch = "main" }
+denokv_sqlite = { git = "https://github.com/denoland/denokv", branch = "main" }
```
Comment (bartlomieju) [inline Cargo.toml:375]: Release `denokv_*` crates before landing

File: ext/fetch/Cargo.toml [hunk 5]
```diff
@@ -19,6 +19,7 @@ data-url.workspace = true
 deno_core.workspace = true
 deno_tls.workspace = true
 dyn-clone = "1"
+http.workspace = true
 http_v02.workspace = true
```
Comment (bartlomieju) [inline ext/fetch/Cargo.toml:23]: Try to remove this one

File: ext/tls/lib.rs [hunk 6]
```diff
@@ -27,14 +29,14 @@ use serde::Deserialize;
...
 use std::sync::Arc;
-use std::time::SystemTime;
 
 mod tls_key;
 pub use tls_key::*;
 
-pub type Certificate = rustls::Certificate;
-pub type PrivateKey = rustls::PrivateKey;
+// pub type Certificate = rustls::pki_types::CertificateDer<'static>;
+// pub type PrivateKey = rustls::pki_types::PrivateKeyDer<'static>;
```
Comment (bartlomieju) [inline ext/tls/lib.rs:39]: Remove

File: ext/tls/lib.rs [hunk 7]
```diff
@@ -149,17 +180,23 @@ pub struct BasicAuth {
...
 
 pub fn create_default_root_cert_store() -> RootCertStore {
+  let root_cert_store = rustls::RootCertStore {
+    roots: webpki_roots::TLS_SERVER_ROOTS.to_vec(),
+  };
+  debug_assert!(!root_cert_store.is_empty());
+  root_cert_store
+}
+
+pub fn create_platform_cert_store() -> RootCertStore {
```
Comment (bartlomieju) [inline ext/tls/lib.rs:190]: This seems not needed, remove.

File: ext/tls/lib.rs [hunk 8]
```diff
@@ -183,7 +220,7 @@ pub fn create_client_config(
 ) -> Result<ClientConfig, AnyError> {
   if let Some(ic_allowlist) = unsafely_ignore_certificate_errors {
     let client_config = ClientConfig::builder()
-      .with_safe_defaults()
+      .dangerous()
```
Comment (bartlomieju) [inline ext/tls/lib.rs:223]: Remove

File: ext/tls/lib.rs [hunk 9]
```diff
@@ -204,18 +241,17 @@ pub fn create_client_config(
...
+  let client_config = ClientConfig::builder().with_root_certificates({
+    let mut root_cert_store =
+      root_cert_store.unwrap_or_else(create_default_root_cert_store);
+    // If custom certs are specified, add them to the store
+    for cert in ca_certs {
+      let reader = &mut BufReader::new(Cursor::new(cert));
+      // This function does not return specific errors, if it fails give a generic message.
+      for r in rustls_pemfile::certs(reader) {
+        match r {
+          Ok(cert) => {
```
Comment (bartlomieju) [inline ext/tls/lib.rs:253]: Hoist out

File: ext/tls/lib.rs [hunk 10]
```diff
@@ -257,15 +294,17 @@ fn add_alpn(client: &mut ClientConfig, socket_use: SocketUse) {
...
+
+  let certs = certs
     .map_err(|_| custom_error("InvalidData", "Unable to decode certificate"))?;
 
   if certs.is_empty() {
     return Err(cert_not_found_err());
   }
 
-  Ok(certs.into_iter().map(rustls::Certificate).collect())
+  Ok(certs.into_iter().map(|x| x.into_owned()).collect())
```
Comment (bartlomieju) [inline ext/tls/lib.rs:307]: This closure might not be needed

File: ext/tls/tls_key.rs [hunk 11]
```diff
@@ -251,13 +257,13 @@ impl TlsKeyLookup {
 pub mod tests {
   use super::*;
   use deno_core::unsync::spawn;
-  use rustls::Certificate;
-  use rustls::PrivateKey;
+  use webpki::types::CertificateDer;
+  use webpki::types::PrivateKeyDer;
```
Comment (bartlomieju) [inline ext/tls/tls_key.rs:261]: This shouldn't be needed.

File: ext/tls/tls_key.rs [hunk 12]
```diff
@@ -251,13 +257,13 @@ impl TlsKeyLookup {
...
 
   fn tls_key_for_test(sni: &str) -> TlsKey {
     TlsKey(
-      vec![Certificate(format!("{sni}-cert").into_bytes())],
-      PrivateKey(format!("{sni}-key").into_bytes()),
+      vec![CertificateDer::from(format!("{sni}-cert").into_bytes())],
+      PrivateKeyDer::try_from(format!("{sni}-key").into_bytes()).unwrap(),
```
Comment (bartlomieju) [inline ext/tls/tls_key.rs:266]: Doesn't work

File: tests/wpt/suite
Comment (bartlomieju) [inline tests/wpt/suite:1]: Revert

File: ext/tls/tls_key.rs (hunk 12 above)
Comment (bartlomieju) [reply ext/tls/tls_key.rs:266]: Probably need to generate some actual proper certificates for these tests

File: ext/tls/Cargo.toml [hunk 13]
```diff
@@ -15,8 +15,8 @@ path = "lib.rs"
 
 [dependencies]
 deno_core.workspace = true
-deno_native_certs = "0.2.0"
-rustls = { workspace = true, features = ["dangerous_configuration"] }
+deno_native_certs = { git = "https://github.com/denoland/deno_native_certs", branch = "upgrade-pemfile2" }
```
Comment (ry) [inline ext/tls/Cargo.toml:18]: release deno_native_certs before landing
https://github.com/denoland/deno_native_certs/pull/3

File: Cargo.lock (hunk 2 above)
Comment (ry) [reply Cargo.lock:3556]: https://github.com/denoland/deno/pull/24108

**Comment by lucacasonato** [comment]
Failing test is due to this change in hyper: https://github.com/seanmonstar/reqwest/pull/2207

My two cents: https://github.com/seanmonstar/reqwest/issues/2240#issuecomment-2151986714

If we land this as is, POST and PUT requests to some HTTP servers, including anything behind Google Load Balancer, probably break. As such I think we're blocked here until this is fixed in reqwest.

**Comment by bartlomieju** [comment]
I have a few failing tests for Node compat:
```
Node.js compatibility ... Node.js compatibility "parallel/test-http-url.parse-auth-with-header-in-request.js" => ./tests/node_compat/test.ts:58:11
Node.js compatibility ... Node.js compatibility "parallel/test-http-url.parse-auth.js" => ./tests/node_compat/test.ts:58:11
Node.js compatibility ... Node.js compatibility "parallel/test-http-url.parse-search.js" => ./tests/node_compat/test.ts:58:11
Node.js compatibility ... Node.js compatibility "parallel/test-http-url.parse-path.js" => ./tests/node_compat/test.ts:58:11
Node.js compatibility ... Node.js compatibility "parallel/test-http-url.parse-post.js" => ./tests/node_compat/test.ts:58:11
Node.js compatibility ... Node.js compatibility "parallel/test-http-url.parse-basic.js" => ./tests/node_compat/test.ts:58:11
```

These tests are writing a response and then closing the server immediately. If the server is not closed immediately, they are passing fine. It appears there's a subtle timing issue introduced with upgrade of `reqwest`.

**Comment by bartlomieju** [comment]
The problem appears to be that `Server.close()` in `node:http` abruptly closes all connections instead of gracefully shutting down the connection and allowing existing requests to complete.

**Comment by bartlomieju** [comment]
These tests were actually broken before, but error handling masked that by ignoring errors that shouldn't be ignored. Since `node:http` will be rewritten in the near future after we migrate off of `reqwest` to `hyper` I will reenable them then.

File: ext/tls/testdata/README [hunk 14]
```diff
@@ -0,0 +1,4 @@
+
+penssl req -x509 -newkey rsa:2048 -nodes -keyout example2_prikey.pem -out example2_cert.der -subj "/C=US/ST=State/L=Locality/O=Organization/CN=example2.com" -outform der
```
Comment (littledivy) [inline ext/tls/testdata/README:2]:
```suggestion
openssl req -x509 -newkey rsa:2048 -nodes -keyout example2_prikey.pem -out example2_cert.der -subj "/C=US/ST=State/L=Locality/O=Organization/CN=example2.com" -outform der
```

File: tests/util/server/src/https.rs [hunk 15]
```diff
@@ -68,30 +68,33 @@ pub fn get_tls_config(
...
-      Some(pkcs8_key[0].clone())
-    } else if !rsa_key.is_empty() {
-      Some(rsa_key[0].clone())
+    let pkcs8_keys_result = rustls_pemfile::pkcs8_private_keys(&mut key_reader)
+      .collect::<Result<Vec<_>, _>>();
+    let pkcs8_keys = pkcs8_keys_result?;
+
+    let rsa_keys_result = rustls_pemfile::rsa_private_keys(&mut key_reader)
+      .collect::<Result<Vec<_>, _>>();
+    let rsa_keys = rsa_keys_result?;
```
Comment (littledivy) [inline tests/util/server/src/https.rs:90]: nit: are there lifetime constraints here to do a let binding?

```suggestion
    let pkcs8_keys = rustls_pemfile::pkcs8_private_keys(&mut key_reader)
      .collect::<Result<Vec<_>, _>>()?;
    let rsa_keys = rustls_pemfile::rsa_private_keys(&mut key_reader)
      .collect::<Result<Vec<_>, _>>()?;
```

**Review by littledivy — APPROVED** [review]
LGTM!

**Review by devsnek — APPROVED** [review]

**Pull request merged**